### PR TITLE
[test] support multiple backbone interfaces in otbr docker test

### PR DESCRIPTION
**Pull request by zesonzhang**
In previous otbr docker tests, when creating docker containers, all the containers(otbr nodes) are connected to the same docker network bridge `backbone0` (when the env `PORT_OFFSET` is not set or set to 0), this means all the otbr nodes are connected to the same infrastructures.

This commit adds support to enable user to config otbr instance infrastructures separately in `TOPOLOGY` when defining test cases , this provides flexibility to run multi-ail related test cases.

The format to define backbone interface per node is:

```
class NewTestCase(thread_cert.TestCase):
    ...
    BR = 1
    ...
    TOPOLOGY = {
        BR: {
            ...
	    'is_otbr': True,
	    'backbone_network': <backbone-name>,
	    ...
	}
	...
    }
    ...
```

`'backbone_network': <backbone-name>` is optional, when it's not given while defining a otbr node, the backbone is default as
`BACKBONE_DOCKER_NETWORK_NAME`. The `<backbone-name>` is suggested to be defined as `backbone[0-9]` to make it more easy to read and understand.

This commit also adds test_multi_ail.py as an example test case to use this new method, this test case checks the two otbr nodes are connected to the different infra and are in the same Thread mesh network.

--------------------
Some TODO items to fully support multiple backbone (sniffer, test report etc) is tracked in https://github.com/openthread/openthread/issues/10551

### CODE REVIEW

File: tests/scripts/thread-cert/border_router/test_multi_ail.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        BR2: {
+            'name': 'BR_2',
+            'backbone': 'backbone1',
+            'allowlist': [BR1],
+            'is_otbr': True,
+            'version': '1.3',
+        }
+    }
+
+    def test(self):
```
Comment (bukepo) [inline tests/scripts/thread-cert/border_router/test_multi_ail.py:72]: Missing steps verifying the two nodes are on different backbone links?

File: tests/scripts/thread-cert/thread_cert.py [hunk 2]
```diff
@@ -522,10 +524,36 @@ def _has_backbone_traffic(self):
         return False
 
     def _prepare_backbone_network(self):
-        network_name = config.BACKBONE_DOCKER_NETWORK_NAME
-        self.assure_run_ok(
-            f'docker network create --driver bridge --ipv6 --subnet {config.BACKBONE_PREFIX} -o "com.docker.network.bridge.name"="{network_name}" {network_name} || true',
-            shell=True)
+        """
```
Comment (bukepo) [inline tests/scripts/thread-cert/thread_cert.py:527]: It looks like `_start_backbone_sniffer` is not updated to support different backbone link.

File: tests/scripts/thread-cert/thread_cert.py [hunk 3]
```diff
@@ -162,14 +162,13 @@ def _setUp(self):
...
-                version=params['version'],
-                is_bbr=params['is_bbr'],
-            )
+            node = nodeclass(i,
+                             is_mtd=params['is_mtd'],
+                             simulator=self.simulator,
+                             name=params.get('name'),
+                             version=params['version'],
+                             is_bbr=params['is_bbr'],
+                             backbone=params['backbone'])
```
[the remark this answers is not in the document]
Comment (zesonzhang) [reply tests/scripts/thread-cert/thread_cert.py:171]: I updated this due it's formatted by "./script/make-pretty python"

File: tests/scripts/thread-cert/node.py [hunk 4]
```diff
@@ -64,8 +64,9 @@ class OtbrDocker:
     _docker_proc = None
     _border_routing_counters = None
 
-    def __init__(self, nodeid: int, **kwargs):
+    def __init__(self, nodeid: int, backbone: str, **kwargs):
         self.verbose = int(float(os.getenv('VERBOSE', 0)))
+        self.backbone = backbone
```
[the remark this answers is not in the document]
Comment (zesonzhang) [reply tests/scripts/thread-cert/node.py]: updated in the latest commit

File: tests/scripts/thread-cert/node.py [hunk 5]
```diff
@@ -64,8 +64,9 @@ class OtbrDocker:
     _docker_proc = None
     _border_routing_counters = None
 
-    def __init__(self, nodeid: int, **kwargs):
+    def __init__(self, nodeid: int, backbone: str, **kwargs):
```
[the remark this answers is not in the document]
Comment (zesonzhang) [reply tests/scripts/thread-cert/node.py]: updated in the latest commit

File: tests/scripts/thread-cert/thread_cert.py (hunk 2 above)
Comment (zesonzhang) [reply tests/scripts/thread-cert/thread_cert.py:527]: Absolutely! Sniffer support is currently limited to backbone0.  We've got enhancements planned to enable multi-backbone support across all cases, which is tracked in https://github.com/openthread/openthread/issues/10551. I intentionally excluded this update from this PR to keep it small, and of course without breaking anything.

File: tests/scripts/thread-cert/border_router/test_multi_ail.py (hunk 1 above)
Comment (zesonzhang) [reply tests/scripts/thread-cert/border_router/test_multi_ail.py:72]: it's in line#86

File: tests/scripts/thread-cert/border_router/test_multi_ail.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+        self.simulator.go(2)
+        br2.start()
+        self.simulator.go(config.BORDER_ROUTER_STARTUP_DELAY)
+
+        # check roles
+        self.assertEqual('leader', br1.get_state())
+        self.assertEqual('router', br2.get_state())
+
+        # check two BRs AIL are in different subnets
+        br1_ifname = br1.bash(IPV4_CIDR_ADDR_CMD)
```
Comment (bukepo) [inline tests/scripts/thread-cert/border_router/test_multi_ail.py:87]: it doesn't look like a ifname, but more like a ip address.

Comment (zesonzhang) [reply tests/scripts/thread-cert/border_router/test_multi_ail.py]: oh, that's absolutely right! updated, thanks!

File: tests/scripts/thread-cert/config.py [hunk 7]
```diff
@@ -64,8 +64,9 @@
 DOMAIN_PREFIX_ALTER = 'fd00:7d04:7d04:7d04::/64'
 
 PORT_OFFSET = int(os.getenv('PORT_OFFSET', '0'))
```
Comment (superwhd) [inline tests/scripts/thread-cert/config.py:66]: I wonder why it's named as 'port'. is it due to a reuse of the env variable?

File: tests/scripts/thread-cert/config.py [hunk 8]
```diff
@@ -64,8 +64,9 @@
...
-BACKBONE_PREFIX = f'{0x9100 + PORT_OFFSET:04x}::/64'
-BACKBONE_PREFIX_REGEX_PATTERN = f'^{0x9100 + PORT_OFFSET:04x}:'
+BACKBONE_IPV6_ADDR_START_BASE = 0x9100
+BACKBONE_PREFIX = f'{BACKBONE_IPV6_ADDR_START_BASE + PORT_OFFSET:04x}::/64'
+BACKBONE_PREFIX_REGEX_PATTERN = f'^{BACKBONE_IPV6_ADDR_START_BASE + PORT_OFFSET:04x}:'
```
Comment (superwhd) [inline tests/scripts/thread-cert/config.py:69]:
```suggestion
BACKBONE_IPV6_ADDR_START_BASE = 0x9100
BACKBONE_IPV6_ADDR_START = BACKBONE_IPV6_ADDR_START_BASE + PORT_OFFSET
BACKBONE_PREFIX = f'{BACKBONE_IPV6_ADDR_START:04x}::/64'
BACKBONE_PREFIX_REGEX_PATTERN = f'^{BACKBONE_IPV6_ADDR_START:04x}:'
```
nit

File: tests/scripts/thread-cert/thread_cert.py [hunk 9]
```diff
@@ -522,10 +524,36 @@ def _has_backbone_traffic(self):
...
+        """
+        Prepares one or multiple backbone network(s).
+
+        This method creates the backbone network based on the `backbone` values defined in the TOPOLOGY in each test.
+        If no backbone values are defined, it sets the default backbone network as BACKBONE_DOCKER_NETWORK_NAME
+        from the config module.
+
+        Args:
+            None
+
```
Comment (superwhd) [inline tests/scripts/thread-cert/thread_cert.py:536]:
```suggestion
```

File: tests/scripts/thread-cert/thread_cert.py [hunk 10]
```diff
@@ -522,10 +524,36 @@ def _has_backbone_traffic(self):
...
+        for node in self.TOPOLOGY:
+            backbone = self.TOPOLOGY[node].get('backbone_network')
+            if backbone:
+                backbone_set.add(backbone)
+
+        # Set default backbone network if backbone_set is empty
+        if not backbone_set:
+            backbone_set.add(config.BACKBONE_DOCKER_NETWORK_NAME)
+
+        # Iterate over the backbone_set and create the backbone network
```
Comment (superwhd) [inline tests/scripts/thread-cert/thread_cert.py:551]:
```suggestion
        # Iterate over the backbone_set and create the backbone networks
```

File: tests/scripts/thread-cert/thread_cert.py [hunk 11]
```diff
@@ -522,10 +524,36 @@ def _has_backbone_traffic(self):
...
+        for offset, backbone in enumerate(backbone_set, start=PORT_OFFSET):
+            backbone_prefix = f'{config.BACKBONE_IPV6_ADDR_START_BASE + offset:04x}::/64'
```
Comment (superwhd) [inline tests/scripts/thread-cert/thread_cert.py:553]: `PORT_OFFSET` is not used here. Is it as intended?

File: tests/scripts/thread-cert/border_router/test_multi_ail.py [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+        }
+    }
+
+    def test(self):
+        br1 = self.nodes[self.BR1]
+        br2 = self.nodes[self.BR2]
+
+        # start nodes
+        br1.start()
+        self.simulator.go(2)
```
Comment (superwhd) [inline tests/scripts/thread-cert/border_router/test_multi_ail.py:78]: Is this needed?

File: tests/scripts/thread-cert/border_router/test_multi_ail.py [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+        self.assertNotEqual(ipaddress.ip_network(br1_infra_ip_addr[0].strip(), strict=False),
+                            ipaddress.ip_network(br2_infra_ip_addr[0].strip(), strict=False))
+
+        # ping each other
+        self.assertTrue(br1.ping(br2.get_ip6_address(config.ADDRESS_TYPE.ML_EID)))
+        self.assertTrue(br2.ping(br1.get_ip6_address(config.ADDRESS_TYPE.ML_EID)))
+
+
+if __name__ == '__main__':
+    unittest.main(verbosity=2)
```
Comment (superwhd) [inline tests/scripts/thread-cert/border_router/test_multi_ail.py:101]: Is the verbosity argument necessary?

File: tests/scripts/thread-cert/config.py (hunk 7 above)
Comment (zesonzhang) [reply tests/scripts/thread-cert/config.py:66]: not sure the history here, but looks like it's related to the test cases parallel running

File: tests/scripts/thread-cert/config.py (hunk 8 above)
Comment (zesonzhang) [reply tests/scripts/thread-cert/config.py:69]: how about removing `BACKBONE_IPV6_ADDR_START_BASE `, directly make
```
BACKBONE_IPV6_ADDR_START = 0x9100  + PORT_OFFSET
```

File: tests/scripts/thread-cert/thread_cert.py (hunk 11 above)
Comment (zesonzhang) [reply tests/scripts/thread-cert/thread_cert.py:553]: `offset` actually will initially assigned as `PORT_OFFSET`

File: tests/scripts/thread-cert/border_router/test_multi_ail.py (hunk 12 above)
Comment (zesonzhang) [reply tests/scripts/thread-cert/border_router/test_multi_ail.py:78]: yes, need to start br1 earlier to ensure it become a leader.

File: tests/scripts/thread-cert/border_router/test_multi_ail.py (hunk 13 above)
Comment (zesonzhang) [reply tests/scripts/thread-cert/border_router/test_multi_ail.py:101]: I prefer to add it to get some more output. 

it actually only have one line more output to show what test case is currently running.

File: tests/scripts/thread-cert/config.py (hunk 8 above)
Comment (superwhd) [reply tests/scripts/thread-cert/config.py:69]: SGTM

**Review by superwhd — APPROVED** [review]
LGTM 👍

**Review by bukepo — APPROVED** [review]

File: tests/scripts/thread-cert/config.py (hunk 8 above)
Comment (zesonzhang) [reply tests/scripts/thread-cert/config.py:69]: emmm, I take a detailed look on the impact for `_prepare_backbone_network()`, looks like `BACKBONE_IPV6_ADDR_START_BASE` is still needed by `_prepare_backbone_network()`. I will go what you have suggested. Thanks!

**Comment by zesonzhang** [comment]
@abtink @jwhui can you please kindly take a look? Thanks!

**Review by abtink — APPROVED** [review]
LGTM. Thanks. 👍

**Review by jwhui — APPROVED** [review]

**Pull request merged**